onnx/defs/shape_inference.h

```diff
@@ -805,4 +913,62 @@ inline void UnionShapeInfo(
   }
 }
 
+// target-type = Union (target-type, source-type)
+// target and source are required to have the same type.
+// Example 1: same tensor type, different shape
+//    source: tensor elem_type: int64, shape: (2, 3, 4, 'x')
+//    target: tensor elem_type: int64, shape: (2, 'y', 5, 'x')
+//    output: tensor elem_type: int64, shape: (2, None, None, 'x')
+// Example 2: same sequence type, different shape
+//    source: sequence of tensor, elem_type: float, shape: (2, 3, 4)
+//    target: sequence of tensor, elem_type: float, shape: None
+//    output: sequence of tensor, elem_type: float, shape: None
+inline void UnionTypeInfo(
+    const TypeProto& source_type,
+    TypeProto& target_type) {
+  if (source_type.value_case() != target_type.value_case()) {
+    fail_type_inference(
+        "Mismatched type:",
+        " source=",
+        source_type.value_case(),
+        " target=",
+        target_type.value_case());
+  }
+
+  if (target_type.has_tensor_type()) {
+    auto source_elem_type = source_type.tensor_type().elem_type();
+    auto target_elem_type = target_type.tensor_type().elem_type();
+
+    if (source_elem_type != target_elem_type) {
+      fail_type_inference(
+          "Mismatched tensor element type:",
+          " source=",
+          source_elem_type,
+          " target=",
+          target_elem_type);
+    }
+
+    UnionShapeInfo(source_type.tensor_type().shape(), *target_type.mutable_tensor_type());
+  } else if (target_type.has_sequence_type()) {
+    if (source_type.sequence_type().elem_type().has_tensor_type() &&
```

Comment:
Better to just call "UnionTypeInfo(source_type.sequence_type().elem_type(), target_type.sequence_type().elem_type())" ... that will handle the case correctly in case one is a sequence of tensors and the other is a sequence of non-tensor-type.

Reply:
thanks! makes sense

Reply:
Thanks for addressing it! I think the nested condition "source_type ... has_tensor_type() && target_type ... has_tensor_type()" should also be removed.

Reply:
Was that meant to check "has_elem_type()"? I assumed a sequence-type will always have an "elem_type". We could check for that to be safe I suppose ... but we should probably treat it as an error if there is no elem_type.